scipy/constants/_constants.py

```diff
@@ -226,7 +225,7 @@
 
 
 def convert_temperature(
-    val: npt.ArrayLike,
+    val: Any,
```

Comment:
Seems reasonable for now. We are waiting for progress on the array API end for type hints.

Reply:
I would just not touch type annotations. Doing so is both unnecessary as long as array API support is experimental, and a regressions for current type annotation users. It'd be better to deal with all type annotations for array-like arguments in one go.

Reply:
Happy to leave alone but wasn't sure how to keep mypy happy in the meantime?

Reply:
Was it actually complaining? We don't run Mypy in the array-api CI job right?

Reply:
Originally I did not change the type hints, see https://github.com/scipy/scipy/actions/runs/8858072080/job/24326283132
![image](https://github.com/scipy/scipy/assets/60778417/8cb569da-76ac-4cbd-a9e7-0e87c458a163)

Reply:
Hmm okay, not immediately clear to me what is going on there. Let me have a closer look. This will be a problem in other places where we have type annotations as well then I suspect.

Reply:
The problem turned out to be fairly straightforward to resolve once I figured out what it was: https://mypy.readthedocs.io/en/stable/common_issues.html#redefinitions-with-incompatible-types. When you do something like:
```
val = _asarray(val, xp=xp, subok=True)
```
and the type of `val` changes in any way other than it becoming more narrow, Mypy gets really unhappy. All the errors in the screenshot above are for Mypy reusing `npt.Arraylike` instead of `np.ndarray` after the change to add array API support. `ndarray` was narrower, `Any` is not.

Unfortunately Mypy gives extremely confusing errors for this, rather than just telling you "hey, this type isn't more narrow, I am reusing the originally declared one".

I pushed a fix that should resolve it. 

Once this is merged, I have type annotations for `_lib/_array_api.py` ready to go, but they're not actually needed for this PR.